### ISSUE TITLE
Could try redrawing an already destroyed wibox

### ISSUE THREAD

**Issue by psychon** [awesomeWM/awesome#520]
I've been thinking about a bug for a few days and I think I finally figured it out.

The symptom can be seen in `tests/test-benchmark.lua`. With the following patch, awesome crashes when the test is run:
```diff
diff --git a/tests/test-benchmark.lua b/tests/test-benchmark.lua
index 065b8e4..ccdad52 100644
--- a/tests/test-benchmark.lua
+++ b/tests/test-benchmark.lua
@@ -66,7 +66,9 @@ local function e2e_tag_switch()
     do_pending_repaint()
 end
 
+benchmark(e2e_tag_switch, "tag switch")
 benchmark(create_and_draw_wibox, "create&draw wibox")
+benchmark(e2e_tag_switch, "tag switch")
 benchmark(update_textclock, "update textclock")
 benchmark(relayout_textclock, "relayout textclock")
 benchmark(redraw_textclock, "redraw textclock")
```
The crash occurs in `cairo_region_union_rectangle` and when you ask valgrind, it tells you that this is a use-after-free. I did not understand this.

Today I somehow came up with https://github.com/pavouk/lgi/issues/122 and then I understood it. The key is the following part from http://www.lua.org/manual/5.2/manual.html#2.5.2:

> Resurrected objects (that is, objects being finalized and objects accessible only through objects being finalized) have a special behavior in weak tables. They are removed from weak values before running their finalizers, but are removed from weak keys only in the next collection after running their finalizers, when such objects are actually freed. This behavior allows the finalizer to access properties associated with the object through weak tables.

The problem comes through `weak_connect_signal`. Internally it uses a weak table and thus the above behavior applies. This means that finalizers for objects which can only be reached via weak signals run before the weak signal is actually disconnected. This in turn means that if the weak signal gets emitted, the callback function can work with userdata which was already finalized. And this explains how the above crash happens.
In the above case this means that changing the tag will cause a taglist to be updated that is only weakly reachable and whose wibox was already finalized.

When lgi gets fixed, the crash doesn't just go away, but instead will likely be replaced with an error like "trying to work on object that was already destroyed" (I know this because I sometimes see this problem when the Lua code manages to use the drawable before some lgi object and then we get such an error).

So... I guess the code behind `weak_connect_signal` needs to be changed to use tables with weak values instead of weak keys? Or does anyone have other ideas?

**Comment by blueyed** [comment]
Your description and solution makes sense to me.
I don't see another way, but do not know much about Lua's internals on the other hand.

**Comment by psychon** [comment]
Only problem with the solution is that it doesn't work and my attempts in the last hour to reproduce the problem with a test case didn't work...

**Comment by psychon** [comment]
The following patch "makes the problem go away", but doesn't really count as a solution to me (but I guess this further confirms the GC-related theory):
```diff
diff --git a/lib/wibox/drawable.lua b/lib/wibox/drawable.lua
index c9aeda0..4ed8e03 100644
--- a/lib/wibox/drawable.lua
+++ b/lib/wibox/drawable.lua
@@ -64,6 +64,10 @@ local function get_widget_context(self)
 end
 
 local function do_redraw(self)
+    if self.gcd then
+        print(string.format("%s was already finalized, aborting redraw (%s)", self, self.drawable))
+        return
+    end
     local surf = surface(self.drawable.surface)
     -- The surface can be nil if the drawable's parent was already finalized
     if not surf then return end
@@ -305,6 +309,10 @@ function drawable.new(d, widget_context_skeleton, drawable_name)
     ret._dirty_area = cairo.Region.create()
     setup_signals(ret)
 
+    ret.gc_check = setmetatable({}, { __gc = function()
+        ret.gcd = true
+    end })
+
     for k, v in pairs(drawable) do
         if type(v) == "function" then
             ret[k] = v
@@ -362,6 +370,10 @@ function drawable.new(d, widget_context_skeleton, drawable_name)
 
     -- Set up our callbacks for repaints
     ret._redraw_callback = function(hierarchy, arg)
+        if ret.gcd then
+            print(string.format("%s was already finalized, ignoring redraw callback (%s)", ret, ret.drawable))
+            return
+        end
         if ret._widget_hierarchy_callback_arg ~= arg then
             return
         end
```

**Comment by psychon** [comment]
Another patch which fixes this (but breaks some unit tests where, for some reason, a weak signal does not get disconnect while it should):
```diff
diff --git a/lib/gears/object.lua b/lib/gears/object.lua
index e255991..ed0e739 100644
--- a/lib/gears/object.lua
+++ b/lib/gears/object.lua
@@ -12,6 +12,17 @@ local error = error
 
 local object = { mt = {} }
 
+local weaks = setmetatable({}, { __mode = "kv" })
+local function intern(f)
+    local res = weaks[f]
+    if not res then
+        res = {}
+        weaks[f] = res
+        weaks[res] = f
+    end
+    return res
+end
+
 --- Verify that obj is indeed a valid object as returned by new()
 local function check(obj)
     if type(obj) ~= "table" or type(obj._signals) ~= "table" then
@@ -40,7 +51,7 @@ function object:add_signal(name)
     if not self._signals[name] then
         self._signals[name] = {
             strong = {},
-            weak = setmetatable({}, { __mode = "k" })
+            weak = {}
         }
     end
 end
@@ -51,7 +62,7 @@ end
 function object:connect_signal(name, func)
     assert(type(func) == "function", "callback must be a function, got: " .. type(func))
     local sig = find_signal(self, name, "connect to")
-    assert(sig.weak[func] == nil, "Trying to connect a strong callback which is already connected weakly")
+    assert(sig.weak[intern(func)] == nil, "Trying to connect a strong callback which is already connected weakly")
     sig.strong[func] = true
 end
 
@@ -63,7 +74,7 @@ function object:weak_connect_signal(name, func)
     assert(type(func) == "function", "callback must be a function, got: " .. type(func))
     local sig = find_signal(self, name, "connect to")
     assert(sig.strong[func] == nil, "Trying to connect a weak callback which is already connected strongly")
-    sig.weak[func] = true
+    sig.weak[intern(func)] = true
 end
 
 --- Disonnect to a signal
@@ -71,7 +82,7 @@ end
 -- @param func The callback that should be disconnected
 function object:disconnect_signal(name, func)
     local sig = find_signal(self, name, "disconnect from")
-    sig.weak[func] = nil
+    sig.weak[intern(func)] = nil
     sig.strong[func] = nil
 end
 
@@ -86,8 +97,17 @@ function object:emit_signal(name, ...)
     for func in pairs(sig.strong) do
         func(self, ...)
     end
-    for func in pairs(sig.weak) do
-        func(self, ...)
+    local clear = {}
+    for handle in pairs(sig.weak) do
+        func = weaks[handle]
+        if func then
+            func(self, ...)
+        else
+            table.insert(clear, handle)
+        end
+    end
+    for _, v in pairs(clear) do
+        sig.weak[v] = nil
     end
 end
 
```

**Comment by psychon** [comment]
New idea for a patch:

Currently we do `weak[func] = true` in `:weak_connect_signal`. Changing this to `weak[func] = func` (and changing the table to `__mode = "kv"` instead of `__mode = "k") should fix the problem (values are removed before they are finalized, keys afterwards; thus this should disconnect the signal before running some finalizers).

This idea is untested, but seems to work with the following test:
```lua
o = object()

do
	local gcable
	local function b()
		print("b", gcable)
	end
	gcable = setmetatable({}, { __gc = function() print("gc'd") b() print("done gc'd") end })
	o:add_signal("")
	o:weak_connect_signal("", b)
end

o:emit_signal("")
collectgarbage("collect")
print("collected")
o:emit_signal("")
```

**Comment by psychon** [comment]
The above does not work with lua 5.1.

So, new idea: Remove `:weak_connect_signal` since its apparently unsafe (as this problem shows). I can replace its only use via "some other magic". That magic would be something like:

```lua
local function gc(obj)
   obj.dead = true
end
local obj = setmetatable({ dead = false }, { __gc = gc })
```
(And then check if `dead == true` where needed and do nothing if that's the case)

TODO: Does `gears.timer.weak_start_signal` suffer from a similar problem (timer runs even though objects were already finalized)? Do we want to do something about this (Remove that function? Apply similar magic)? @blueyed What do you think?

**Comment by psychon** [comment]
Yay, here is a version that also works with Lua 5.1 (based on my previous idea of making object._signals.weak contain both weak keys and values and because Lua 5.1 does what we want if the contained value is a userdata instead of a function):

```lua
local make_weak
if _VERSION <= "Lua 5.1" then
    function make_weak(obj)
        local udata = newproxy(true)
        local mt = getmetatable(udata)
        -- Needs a __gc metamethod, else 
        mt.__gc = function() end
        mt.obj = obj
        return udata
    end
else
    function make_weak(obj)
        return obj
    end
end

--- Connect to a signal weakly. This allows the callback function to be garbage
-- collected and automatically disconnects the signal when that happens.
-- @param name The name of the signal
-- @param func The callback to call when the signal is emitted
function object:weak_connect_signal(name, func)
    assert(type(func) == "function", "callback must be a function, got: " .. type(func))
    local sig = find_signal(self, name, "connect to")
    assert(sig.strong[func] == nil, "Trying to connect a weak callback which is already connected strongly")
    sig.weak[func] = make_weak(func)
end
```

**Comment by psychon** [comment]
Should be fixed by #567
